src/Security/RolesMatrixBuilder.php

```diff
@@ -0,0 +1,309 @@
+        return $roles;
+    }
+
+    public function getCustomRolesForView(): array
+    {
+        $roles = [];
+
+        // get roles from the service container
+        foreach ($this->rolesHierarchy as $name => $rolesHierarchy) {
+            $roles[$name] = [
+                'read_only' => !$this->authorizationChecker->isGranted($name),
+            ];
+
+            foreach ($rolesHierarchy as $role) {
+                if (false === array_key_exists($role, $this->rolesHierarchy)
+                    && !isset($roles[$role])
+                    && false === $this->recursiveArraySearch($role, $roles)
+                ) {
+                    $roles[$role] = [
+                        'read_only' => !$this->authorizationChecker->isGranted($role),
+                    ];
+                }
+            }
+        }
+
+        return $roles;
+    }
+
+    public function getExclude(): array
+    {
+        return $this->exclude;
+    }
+
+    public function addExclude(string $exclude): void
+    {
+        $this->exclude[] = $exclude;
+    }
+
+    public function getLabelPermission(): array
+    {
+        return $this->labelPermission;
+    }
+
+    public function getLabelAdmin(): array
+    {
+        return $this->labelAdmin;
+    }
+
+    public function getAdminRolesForView(): array
+    {
+        $roles = [];
+        // get roles from the Admin classes
+        foreach ($this->pool->getAdminServiceIds() as $id) {
+            try {
+                $admin = $this->pool->getInstance($id);
+            } catch (\Exception $e) {
+                $this->logger->error($e->getMessage(), ['exception' => $e]);
+                continue;
+            }
+
+            if (in_array($id, $this->exclude)) {
+                continue;
+            }
+
+            $isMaster = ($admin->isGranted('MASTER') || $admin->isGranted('OPERATOR') ||
+                $this->authorizationChecker->isGranted($this->pool->getOption('role_super_admin')));
+
+            $securityHandler = $admin->getSecurityHandler();
+            $baseRole = $securityHandler->getBaseRole($admin);
+            $groupPermission = $admin->getSecurityInformation();
+            $this->labelPermission = array_keys($groupPermission);
+            $this->labelAdmin[] = $admin->getTranslator()->trans($admin->getLabel());
+
+            if (0 == strlen($baseRole)) { // the security handler related to the admin does not provide a valid string
+                continue;
+            }
+
+            $roles[$baseRole] = [
+                'label' => $admin->getTranslator()->trans($admin->getLabel()),
+            ];
+
+            foreach ($groupPermission as $name => $item) {
+                $role = sprintf($baseRole, $name);
+                $roles[$baseRole]['permissions'][$name] =
+                    !$isMaster && !$this->authorizationChecker->isGranted($role);
+            }
+        }
+
+        return $roles;
+    }
+
+    private function iterateAdminRoles(callable $func): void
```

Comment:
I really don't like that `getAdminRolesForView` and this method are almost the same, is there any way to extract stuff into additional private methods? I mean difference is in 2 lines between methods. If not, just ignore